builder/pipeline.h

```diff
@@ -176,78 +176,56 @@ class func {
   }
   const std::optional<loop_id>& compute_at() const { return compute_at_; }
 
-  // TODO(https://github.com/dsharlet/slinky/issues/8): Try to do this with a variadic template implementation.
-  template <typename Out1>
-  static func make(callable_wrapper<Out1> impl, output out1) {
-    symbol_id out1_sym = out1.sym();
-    return func(
-        [=, impl = std::move(impl)](eval_context& ctx) -> index_t {
-          const raw_buffer* out1_buf = ctx.lookup_buffer(out1_sym);
-          return impl(out1_buf->cast<Out1>());
-        },
-        {}, {std::move(out1)});
-  }
-
-  template <typename In1, typename Out1>
-  static func make(callable_wrapper<const In1, Out1> impl, input in1, output out1) {
-    symbol_id in1_sym = in1.sym();
-    symbol_id out1_sym = out1.sym();
-    return func(
-        [=, impl = std::move(impl)](eval_context& ctx) -> index_t {
-          const raw_buffer* in1_buf = ctx.lookup_buffer(in1_sym);
-          const raw_buffer* out1_buf = ctx.lookup_buffer(out1_sym);
-          return impl(in1_buf->cast<const In1>(), out1_buf->cast<Out1>());
-        },
-        {std::move(in1)}, {std::move(out1)});
-  }
-
-  template <typename In1, typename In2, typename Out1>
-  static func make(callable_wrapper<const In1, const In2, Out1> impl, input in1, input in2, output out1) {
-    symbol_id in1_sym = in1.sym();
-    symbol_id in2_sym = in2.sym();
-    symbol_id out1_sym = out1.sym();
-    return func(
-        [=, impl = std::move(impl)](eval_context& ctx) -> index_t {
-          const raw_buffer* in1_buf = ctx.lookup_buffer(in1_sym);
-          const raw_buffer* in2_buf = ctx.lookup_buffer(in2_sym);
-          const raw_buffer* out1_buf = ctx.lookup_buffer(out1_sym);
-          return impl(in1_buf->cast<const In1>(), in2_buf->cast<const In2>(), out1_buf->cast<Out1>());
-        },
-        {std::move(in1), std::move(in2)}, {std::move(out1)});
+private:
+  template <typename First, typename... Rest>
+  static auto build_tuple(eval_context& ctx, const symbol_id* symbols, std::size_t index = 0) {
+    if constexpr (sizeof...(Rest) == 0) {
```

Comment:
I think this would work and might be better depending on your taste (avoids the recursion to unpack the symbols):

```
template <typename... Ts, std::size_t... Is>
static auto build_tuple(eval_context& ctx, const symbol_id* symbols, std::integer_sequence<Is...>) {
  return std::tuple<const buffer<Ts...>&>(ctx.lookup_buffer(symbols[Is])->template cast<Ts>()...);
}
```

Call site would need to pass `std::make_integer_sequence<Size>()` for the last argument.

You might also be able to just make a `call_impl` helper using this technique and skip the tuple/`std::apply`.

Reply:
The recursion is all unwound at compile time, so the code generated is likely equivalent. I'll try using that instead if you prefer.

Reply:
I understand the generated code is likely equivalent, but I think the code would be easier to understand if we just called it directly. The "added noise" is minimal vs. making a tuple.

Reply:
Ah, I remember... the gotcha with using index_sequence is that we have to expand two parameter packs in tandem, and this makes the compiler unhappy:

```
./builder/pipeline.h:191:36: error: template argument for non-type template parameter must be an expression
    return std::tuple<const buffer<Ts...>&>(ctx.lookup_buffer(symbols[Is])->template cast<Ts>()...);
```

Reply:
You wouldn't have that problem if you were calling the impl directly though?

Reply:
I'll give it a try, but I don't see how that avoids the need to unpack them in tandem.

Reply:
Oh, I see what you mean. You'd have to do this:
```
template <typename... Ts, std::size_t... Is>
static auto build_tuple(eval_context& ctx, const symbol_id* symbols, std::integer_sequence<Is...>) {
  using Types = std::tuple<Ts...>;
  return std::tuple<const buffer<Ts...>&>(ctx.lookup_buffer(symbols[Is])->template cast<std::tuple_element<Is, Types>>()...);
}

Reply:
Ahhh, clever, I'll try it